File: ios/SmashCore/Sources/SmashCore/Match/MatchVector.swift
/// A golden vector of the simulation (spec §4.7): a setup, the player's input as (tick, finger)
/// events, and the match it produces — sampled state every `every` ticks and every event — as text
/// with every double written as the hex of its bits. Android's `MatchVector` reads and writes the
/// same format; both replay shared/vectors/match/ and compare line by line, never with a tolerance.
///
///     setup match seed <hex> sport <id> period <bits> orbit <bits> cup <0|1> control <player|automatic>
///           home <rating> <formation> <6 × tactic bits> away <rating> <formation> <6 × tactic bits>
///     setup drill <id> seed <hex> orbit <bits> tactics <6 × tactic bits>
///     every <N>
///     ticks <most ticks to run>
///     input <tick> down|up        — applied at the boundary once <tick> ticks have run
///     e <tick> <event…>           — what was emitted during the tick that brought the count to <tick>
///     s <tick> <state> <period> <overtime> <clock> <score0> <score1> <stream> <ball x z vx vz>
///       <carrier|-1> <orbit> <orbitDirection> then each player's <x z vx vz facing>, in roster order
public struct MatchVector: Sendable {
    public enum Setup: Sendable {
        case match(MatchSetup)
        case drill(DrillSetup)
    }

    public var setup: Setup
    public var every: Int
    public var maxTicks: Int
    public var inputs: [(tick: Int, down: Bool)]

    public init(setup: Setup, every: Int, maxTicks: Int, inputs: [(tick: Int, down: Bool)]) {
        self.setup = setup; self.every = every; self.maxTicks = maxTicks; self.inputs = inputs
    }

    public func makeMatch() -> Match {
        switch setup {
        case .match(let s): Match(s)
        case .drill(let s): Match(s)
        }
    }

    // MARK: Running

    /// The body lines this vector produces: events and samples, in order.
    public func run() -> [String] {
        var match = makeMatch()
        var lines = match.drainEvents().map { "e 0 " + MatchVector.format($0) }
        lines.append(MatchVector.sample(match))
        var next = 0
        while match.ticks < maxTicks && match.state != .ended {
            while next < inputs.count && inputs[next].tick == match.ticks {
                match.hold(inputs[next].down)
                next += 1
            }
            match.tick()
            for event in match.drainEvents() { lines.append("e \(match.ticks) " + MatchVector.format(event)) }
            if match.ticks % every == 0 || match.state == .ended { lines.append(MatchVector.sample(match)) }
        }
        return lines
    }

    public static func sample(_ m: Match) -> String {
        var t = ["s", String(m.ticks), m.state.rawValue, String(m.period), m.overtime ? "1" : "0", hex(m.clock),
                 String(m.score[0]), String(m.score[1]), hex(m.streamState),
                 hex(m.ball.pos.x), hex(m.ball.pos.z), hex(m.ball.vel.x), hex(m.ball.vel.z),
                 String(m.ball.carrier ?? -1), hex(m.ball.orbit), hex(m.ball.orbitDirection)]
        for p in m.players { t += [hex(p.pos.x), hex(p.pos.z), hex(p.vel.x), hex(p.vel.z), hex(p.facing)] }
        return t.joined(separator: " ")
    }

    public static func format(_ e: MatchEvent) -> String {
        func who(_ i: Int?) -> String { i.map(String.init) ?? "-" }
        switch e {
        case .faceOff(let spot): return "faceoff \(hex(spot.x)) \(hex(spot.z))"
        case .play: return "play"
        case .ready: return "ready"
        case .pickup(let p): return "pickup \(p)"
        case .pass(let from, let to): return "pass \(from) \(to)"
        case .shot(let by, let kind): return "shot \(by) \(kind.rawValue)"
        case .steal(let by, let from): return "steal \(by) \(from)"
        case .save(let by): return "save \(by)"
        case .block(let by): return "block \(by)"
        case .post: return "post"
        case .board(let speed): return "board \(hex(speed))"
        case .goal(let team, let scorer, let assist, let own): return "goal \(team) \(who(scorer)) \(who(assist)) \(own ? 1 : 0)"
        case .whistle: return "whistle"
        case .offside(let team, let player): return "offside \(team) \(player)"
        case .periodEnd(let n): return "periodEnd \(n)"
        case .drillInterrupted(let reason): return "interrupted \(reason.rawValue)"
        case .end(let result): return "end \(result.rawValue)"
        }
    }

    // MARK: The file

    /// The header lines (setup, sampling, input) that precede the body.
    public func headerLines() -> [String] {
        var lines: [String]
        switch setup {
        case .match(let s):
            lines = ["setup match seed \(MatchVector.hex(s.seed)) sport \(s.sport.rawValue) period \(MatchVector.hex(s.periodSeconds)) "
                     + "orbit \(MatchVector.hex(s.orbitPeriod)) cup \(s.cup ? 1 : 0) control \(s.control.rawValue) "
                     + "home \(MatchVector.side(s.home)) away \(MatchVector.side(s.away))"]
        case .drill(let s):
            lines = ["setup drill \(s.drill.rawValue) seed \(MatchVector.hex(s.seed)) orbit \(MatchVector.hex(s.orbitPeriod)) "
                     + "tactics \(MatchVector.tactics(s.tactics))"]
        }
        lines.append("every \(every)")
        lines.append("ticks \(maxTicks)")
        lines += inputs.map { "input \($0.tick) \($0.down ? "down" : "up")" }
        return lines
    }

    /// Parses a vector file: the vector, and the body lines it recorded.
    public static func parse(_ text: String) throws -> (vector: MatchVector, recorded: [String]) {
        var setup: Setup?
        var every: Int?
        var maxTicks: Int?
        var inputs: [(tick: Int, down: Bool)] = []
        var body: [String] = []
        for raw in text.split(separator: "\n", omittingEmptySubsequences: true) {
            let line = String(raw)
            if line.hasPrefix("#") { continue }
            var tokens = Tokens(line.split(separator: " ").map(String.init))
            switch try tokens.word() {
            case "setup": setup = try parseSetup(&tokens)
            case "every": every = try tokens.int()
            case "ticks": maxTicks = try tokens.int()
            case "input":
                let tick = try tokens.int()
                let edge = try tokens.word()
                guard edge == "down" || edge == "up" else { throw VectorFormatError("input edge \(edge) is not down|up") }
                inputs.append((tick, edge == "down"))
            case "s", "e": body.append(line)
            case let other: throw VectorFormatError("unknown line kind \(other)")
            }
        }
        guard let setup, let every, let maxTicks else { throw VectorFormatError("a vector needs setup, every and ticks lines") }
        return (MatchVector(setup: setup, every: every, maxTicks: maxTicks, inputs: inputs), body)
    }

    static func parseSetup(_ t: inout Tokens) throws -> Setup {
        switch try t.word() {
        case "match":
            try t.expect("seed"); let seed = try t.bits()
            try t.expect("sport"); let sportId = try t.word()
            guard let sport = Sport(rawValue: sportId) else { throw VectorFormatError("unknown sport \(sportId)") }
            try t.expect("period"); let period = try t.double()
            try t.expect("orbit"); let orbit = try t.double()
            try t.expect("cup"); let cup = try t.int() == 1
            try t.expect("control"); let controlId = try t.word()
            guard let control = Control(rawValue: controlId) else { throw VectorFormatError("unknown control \(controlId)") }
            try t.expect("home"); let home = try parseSide(&t)
            try t.expect("away"); let away = try parseSide(&t)
            return .match(MatchSetup(seed: seed, sport: sport, home: home, away: away, periodSeconds: period,
                                     orbitPeriod: orbit, cup: cup, control: control))
        case "drill":
            let id = try t.word()
            guard let drill = Drill(rawValue: id) else { throw VectorFormatError("unknown drill \(id)") }
            try t.expect("seed"); let seed = try t.bits()
            try t.expect("orbit"); let orbit = try t.double()
            try t.expect("tactics"); let tactics = try parseTactics(&t)
            return .drill(DrillSetup(drill: drill, seed: seed, tactics: tactics, orbitPeriod: orbit))
        case let other:
            throw VectorFormatError("unknown setup kind \(other)")
        }
    }

    static func parseSide(_ t: inout Tokens) throws -> SideSetup {
        let rating = try t.int()
        let id = try t.word()
        guard let formation = Formation(rawValue: id) else { throw VectorFormatError("unknown formation \(id)") }
        return SideSetup(rating: rating, tactics: try parseTactics(&t), formation: formation)
    }

    static func parseTactics(_ t: inout Tokens) throws -> Tactics {
        Tactics(pressing: try t.double(), covering: try t.double(), pushUp: try t.double(),
                passing: try t.double(), shooting: try t.double(), discipline: try t.double())
    }

    static func side(_ s: SideSetup) -> String { "\(s.rating) \(s.formation.rawValue) \(tactics(s.tactics))" }

    static func tactics(_ t: Tactics) -> String {
        [t.pressing, t.covering, t.pushUp, t.passing, t.shooting, t.discipline].map(hex).joined(separator: " ")
    }

    public static func hex(_ v: UInt64) -> String {
        let s = String(v, radix: 16, uppercase: true)
        return String(repeating: "0", count: 16 - s.count) + s
    }

    public static func hex(_ d: Double) -> String { hex(d.bitPattern) }

    struct Tokens {
        var items: [String]
        var at = 0
        init(_ items: [String]) { self.items = items }

        mutating func word() throws -> String {
            guard at < items.count else { throw VectorFormatError("line ends early: \(items.joined(separator: " "))") }
            defer { at += 1 }
            return items[at]
        }

        mutating func expect(_ w: String) throws {
            let got = try word()
            guard got == w else { throw VectorFormatError("expected \(w), got \(got)") }
        }

        mutating func int() throws -> Int {
            let w = try word()
            guard let v = Int(w) else { throw VectorFormatError("\(w) is not an integer") }
            return v
        }

        mutating func bits() throws -> UInt64 {
            let w = try word()
            guard w.count == 16, let v = UInt64(w, radix: 16) else { throw VectorFormatError("\(w) is not 16 hex digits") }
            return v
        }

        mutating func double() throws -> Double { Double(bitPattern: try bits()) }
    }
}

public struct VectorFormatError: Error, CustomStringConvertible {
    public let description: String
    public init(_ description: String) { self.description = description }
}

File: android/core/src/main/kotlin/in/nann/smashhockey/core/match/MatchVector.kt
package `in`.nann.smashhockey.core.match

import `in`.nann.smashhockey.core.generated.Drill
import `in`.nann.smashhockey.core.generated.Formation
import `in`.nann.smashhockey.core.generated.Sport
import `in`.nann.smashhockey.core.generated.Tactics

/**
 * A golden vector of the simulation (spec §4.7): a setup, the player's input as (tick, finger)
 * events, and the match it produces — sampled state every [every] ticks and every event — as text
 * with every double written as the hex of its bits. iOS's `MatchVector` reads and writes the same
 * format (its doc comment spells the lines out); both replay shared/vectors/match/ line by line,
 * never with a tolerance.
 */
class MatchVector(val setup: Setup, val every: Int, val maxTicks: Int, val inputs: List<Input>) {
    sealed interface Setup {
        data class OfMatch(val setup: MatchSetup) : Setup
        data class OfDrill(val setup: DrillSetup) : Setup
    }

    /** The finger going [down] (or up) at the boundary once [tick] ticks have run. */
    data class Input(val tick: Int, val down: Boolean)

    fun makeMatch(): Match = when (setup) {
        is Setup.OfMatch -> Match(setup.setup)
        is Setup.OfDrill -> Match(setup.setup)
    }

    /** The body lines this vector produces: events and samples, in order. */
    fun run(): List<String> {
        val match = makeMatch()
        val lines = ArrayList<String>()
        for (e in match.drainEvents()) lines.add("e 0 " + format(e))
        lines.add(sample(match))
        var next = 0
        while (match.ticks < maxTicks && match.state != MatchState.ENDED) {
            while (next < inputs.size && inputs[next].tick == match.ticks) {
                match.hold(inputs[next].down)
                next += 1
            }
            match.tick()
            for (e in match.drainEvents()) lines.add("e ${match.ticks} " + format(e))
            if (match.ticks % every == 0 || match.state == MatchState.ENDED) lines.add(sample(match))
        }
        return lines
    }

    companion object {
        fun sample(m: Match): String {
            val b = m.ball
            val t = arrayListOf(
                "s", m.ticks.toString(), m.state.key, m.period.toString(), if (m.overtime) "1" else "0", hex(m.clock),
                m.score[0].toString(), m.score[1].toString(), hex(m.streamState),
                hex(b.pos.x), hex(b.pos.z), hex(b.vel.x), hex(b.vel.z), (b.carrier ?: -1).toString(),
                hex(b.orbit), hex(b.orbitDirection),
            )
            for (p in m.players) {
                t += listOf(hex(p.pos.x), hex(p.pos.z), hex(p.vel.x), hex(p.vel.z), hex(p.facing))
            }
            return t.joinToString(" ")
        }

        fun format(e: MatchEvent): String {
            fun who(i: Int?) = i?.toString() ?: "-"
            return when (e) {
                is MatchEvent.FaceOff -> "faceoff ${hex(e.spot.x)} ${hex(e.spot.z)}"
                MatchEvent.Play -> "play"
                MatchEvent.Ready -> "ready"
                is MatchEvent.Pickup -> "pickup ${e.player}"
                is MatchEvent.Pass -> "pass ${e.from} ${e.to}"
                is MatchEvent.Shot -> "shot ${e.by} ${e.kind.key}"
                is MatchEvent.Steal -> "steal ${e.by} ${e.from}"
                is MatchEvent.Save -> "save ${e.by}"
                is MatchEvent.Block -> "block ${e.by}"
                MatchEvent.Post -> "post"
                is MatchEvent.Board -> "board ${hex(e.speed)}"
                is MatchEvent.Goal -> "goal ${e.team} ${who(e.scorer)} ${who(e.assist)} ${if (e.ownGoal) 1 else 0}"
                MatchEvent.Whistle -> "whistle"
                is MatchEvent.Offside -> "offside ${e.team} ${e.player}"
                is MatchEvent.PeriodEnd -> "periodEnd ${e.period}"
                is MatchEvent.DrillInterrupted -> "interrupted ${e.reason.key}"
                is MatchEvent.End -> "end ${e.result.key}"
            }
        }

        fun hex(bits: Long): String = java.lang.Long.toHexString(bits).uppercase().padStart(16, '0')

        fun hex(d: Double): String = hex(d.toRawBits())

        /** Parses a vector file into the vector and the body lines it recorded. */
        fun parse(text: String): Pair<MatchVector, List<String>> {
            var setup: Setup? = null
            var every: Int? = null
            var maxTicks: Int? = null
            val inputs = ArrayList<Input>()
            val body = ArrayList<String>()
            for (line in text.split("\n")) {
                if (line.isEmpty() || line.startsWith("#")) continue
                val t = Tokens(line.split(" "))
                when (val kind = t.word()) {
                    "setup" -> setup = parseSetup(t)
                    "every" -> every = t.int()
                    "ticks" -> maxTicks = t.int()
                    "input" -> {
                        val tick = t.int()
                        val edge = t.word()
                        require(edge == "down" || edge == "up") { "input edge $edge is not down|up" }
                        inputs.add(Input(tick, edge == "down"))
                    }
                    "s", "e" -> body.add(line)
                    else -> throw IllegalArgumentException("unknown line kind $kind")
                }
            }
            requireNotNull(setup) { "a vector needs a setup line" }
            requireNotNull(every) { "a vector needs an every line" }
            requireNotNull(maxTicks) { "a vector needs a ticks line" }
            return Pair(MatchVector(setup, every, maxTicks, inputs), body)
        }

        private fun parseSetup(t: Tokens): Setup = when (val kind = t.word()) {
            "match" -> {
                t.expect("seed"); val seed = t.bits()
                t.expect("sport"); val sport = Sport.of(t.word())
                t.expect("period"); val period = t.double()
                t.expect("orbit"); val orbit = t.double()
                t.expect("cup"); val cup = t.int() == 1
                t.expect("control"); val control = Control.of(t.word())
                t.expect("home"); val home = parseSide(t)
                t.expect("away"); val away = parseSide(t)
                Setup.OfMatch(MatchSetup(seed, sport, home, away, period, orbit, cup, control))
            }
            "drill" -> {
                val id = t.word()
                val drill = Drill.entries.firstOrNull { it.key == id } ?: throw IllegalArgumentException("unknown drill $id")
                t.expect("seed"); val seed = t.bits()
                t.expect("orbit"); val orbit = t.double()
                t.expect("tactics"); val tactics = parseTactics(t)
                Setup.OfDrill(DrillSetup(drill, seed, tactics, orbit))
            }
            else -> throw IllegalArgumentException("unknown setup kind $kind")
        }

        private fun parseSide(t: Tokens): SideSetup {
            val rating = t.int()
            val formation = Formation.of(t.word())
            return SideSetup(rating, parseTactics(t), formation)
        }

        private fun parseTactics(t: Tokens) = Tactics(
            pressing = t.double(), covering = t.double(), pushUp = t.double(),
            passing = t.double(), shooting = t.double(), discipline = t.double(),
        )
    }

    private class Tokens(val items: List<String>) {
        var at = 0

        fun word(): String {
            require(at < items.size) { "line ends early: ${items.joinToString(" ")}" }
            return items[at++]
        }

        fun expect(w: String) {
            val got = word()
            require(got == w) { "expected $w, got $got" }
        }

        fun int(): Int = word().toInt()

        fun bits(): Long {
            val w = word()
            require(w.length == 16) { "$w is not 16 hex digits" }
            return java.lang.Long.parseUnsignedLong(w, 16)
        }

        fun double(): Double = Double.fromBits(bits())
    }
}
